x/contracts/examples/multisig/src/lib.rs

```diff
@@ -0,0 +1,468 @@
+use std::{collections::HashSet, num::NonZeroUsize};
+
+use wasmlanche::{
+    borsh::{BorshDeserialize, BorshSerialize},
+    public, state_schema, Address, Context,
+};
+
+type ProposalId = usize;
+
+// TODO: make a state_keys module so I can reuse `Proposal`.
+state_schema! {
+    ProposalCount => ProposalId,
+    ProposalItem(ProposalId) => Proposal,
+    ProposalVoteInfo(ProposalId) => ProposalMeta,
+    Vote(ProposalId, Address) => bool,
+}
+
+#[public]
+pub fn propose(ctx: &mut Context, proposal: Proposal, meta: ProposalMeta) -> ProposalId {
+    let actor = ctx.actor();
+
+    let id = ctx
+        .get(ProposalCount)
+        .expect("failed to deserialize")
+        .unwrap_or_default();
+
+    ctx.store((
+        (ProposalCount, id + 1),
+        (ProposalItem(id), proposal),
+        (ProposalVoteInfo(id), meta),
+        (Vote(id, actor), true),
+    ))
+    .expect("failed to store proposal");
+
+    id
+}
+
+#[public]
+pub fn vote(ctx: &mut Context, proposal_id: ProposalId, vote: bool) -> Option<Vec<u8>> {
+    let actor = ctx.actor();
+
+    ctx.store_by_key(Vote(proposal_id, actor), vote)
+        .expect("failed to serialize");
+
+    if !vote {
+        return None;
+    }
+
+    let ProposalMeta {
+        voters,
+        votes_remaining,
+    } = ctx
+        .get(ProposalVoteInfo(proposal_id))
+        .expect("failed to deserialize")
+        .expect("proposal not found");
+
+    if !voters.contains(actor) || votes_remaining == 0 {
+        return None;
+    }
```

Comment:
I'll have to look again, but I think I fixed this

Reply:
Honestly, the reason I did it this way is that it's the voter that's paying for the vote, it's really up to them to ensure whether or not they can vote. 

The logic flows like this
 - if the voter hasn't voted, cast their vote
 - if the vote affects quorum, record the change
 - if a quorum is reached, execute the proposal
 
 Doing things in this order is much simpler as you only deal with one piece of state at a time. Otherwise, there's more branching logic that I just don't care to optimize for.

Reply:
Makes sense! It's just a code smell, for example if someone wants to extend the code and expects the votes in the `Vote` state keys to be part of the voters in order to sum the amount of votes then they might get surprised.

You might as well just write
```suggestion
    let actor = ctx.actor();

    let ProposalMeta {
        voters,
        votes_remaining,
    } = ctx
        .get(ProposalVoteInfo(proposal_id))
        .expect("failed to deserialize")
        .expect("proposal not found");

    if !voters.contains(actor) || votes_remaining == 0 {
        return None;
    }

    ctx.store_by_key(Vote(proposal_id, actor), vote)
        .expect("failed to serialize");

    if !vote {
        return None;
    }
```

The suggestion diff is not perfect but essentially just moving the "effects" after the "checks", but no changes are required.